Internal Fault

Hey Guys, 
After successfully setting up my Minecraft 1.12.2 server on my raspberry pi 4 2gb i run into some problems.
The biggest Problem is, that the server keeps crashing in between every 7minutes to 4 hours.
The output says this:
cuberite has encountered an error and needs to close| sigabrt: server self-terminated due to an internal fault | | cuberite local build (id: unknown) | from commit approx: abc96bb4ce0bfb5627e74f4f2e66f19eaaa12736
We tried loading less Chunks, but the Situation didn't get any better.
I would be really happy, if someone could help me, because we are really enjoying playing minecraft on Cuberite.
Thanks

Thank you, for your help, but this doesn`t seem to work, did i miss a step?
pi@raspberrypi:~/Server $ cmake .. -DCMAKE_BUILD_TYPE=RelWithDebInfo
CMake Error: The source directory "/home/pi" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.
I am really sorry, but i am an absolut amateur....

i think it worked so far....
But i can`t find gdb.txt, dump.core or Cuberite binary.

Here is what i did:
pi@raspberrypi:~/cuberite $ cd Build
pi@raspberrypi:~/cuberite/Build $ cmake .. -DCMAKE_BUILD_TYPE=RelWithDebInfo
-- Version: TRUE7.0.0
-- Build type: RelWithDebInfo
-- CXX_STANDARD: 17
-- Required features: cxx_variadic_templates
-- JsonCpp Version: 1.9.3
--
--         ---( Libevent 2.1.12-stable )---
--
-- Available event backends:
-- CMAKE_BINARY_DIR:         /home/pi/cuberite
-- CMAKE_CURRENT_BINARY_DIR: /home/pi/cuberite/lib/libevent
-- CMAKE_SOURCE_DIR:         /home/pi/cuberite
-- CMAKE_CURRENT_SOURCE_DIR: /home/pi/cuberite/lib/libevent
-- PROJECT_BINARY_DIR:       /home/pi/cuberite/lib/libevent
-- PROJECT_SOURCE_DIR:       /home/pi/cuberite/lib/libevent
-- CMAKE_MODULE_PATH:        /home/pi/cuberite/lib/libevent/cmake/
-- CMAKE_COMMAND:            /usr/bin/cmake
-- CMAKE_ROOT:               /usr/share/cmake-3.13
-- CMAKE_SYSTEM:             Linux-5.4.79-v7l+
-- CMAKE_SYSTEM_NAME:        Linux
-- CMAKE_SYSTEM_VERSION:     5.4.79-v7l+
-- CMAKE_SYSTEM_PROCESSOR:   armv7l
-- CMAKE_SKIP_RPATH:         NO
-- CMAKE_VERBOSE_MAKEFILE:   FALSE
-- CMAKE_C_FLAGS:             -Wall -Wextra -Wno-unused-parameter -Wstrict-alias                                                                                                             ing -Wstrict-prototypes -fno-strict-aliasing -Wmissing-prototypes -Winit-self -W                                                                                                             missing-field-initializers -Wdeclaration-after-statement -Waddress -Wnormalized=                                                                                                             id -Woverride-init -Wlogical-op -Wwrite-strings
-- CMAKE_BUILD_TYPE:         RelWithDebInfo
-- CMAKE_C_COMPILER:         /usr/bin/cc (id GNU, clang 0, GNUC 1)
-- CMAKE_AR:                 /usr/bin/ar
-- CMAKE_RANLIB:             /usr/bin/ranlib
--
-- Using c++ standard c++17
-- CMake version: 3.13.4
-- Project version: 3.1.1
-- CMAKE_CXX_COMPILER '/usr/bin/c++' 'GNU' '8.3.0'
-- CMAKE_CXX_FLAGS                ' -Wall -Wextra -Wpedantic -Wswitch-enum -Wsha                                                                                                             dow -Wno-long-long'
-- CMAKE_CXX_FLAGS_RELWITHDEBINFO '-O2 -g -DNDEBUG'
-- Compile sqlite3 from source in subdirectory
-- SQLITECPP_RUN_CPPLINT OFF
-- SQLITECPP_RUN_CPPCHECK OFF
-- SQLITECPP_RUN_DOXYGEN OFF
-- SQLITECPP_BUILD_EXAMPLES OFF
-- SQLITECPP_BUILD_TESTS OFF
Creating output folder and symlinks...
Creating symlink, orig = /home/pi/cuberite/src/../Server/Install; link = /home/p                                                                                                             i/cuberite/Server/Install
CMake Error at src/CMakeLists.txt:232 (message):
  Could not create symbolic link for: /home/pi/cuberite/Server/Install -->
  /home/pi/cuberite/src/../Server/Install: ln: failed to create symbolic link
  '/home/pi/cuberite/Server/Install': File exists


Call Stack (most recent call first):
  src/CMakeLists.txt:258 (make_symlink)


-- Configuring incomplete, errors occurred!
See also "/home/pi/cuberite/CMakeFiles/CMakeOutput.log".
See also "/home/pi/cuberite/CMakeFiles/CMakeError.log".
pi@raspberrypi:~ $ cd Server
pi@raspberrypi:~/Server $ gdb Cuberite
GNU gdb (Raspbian 8.2.1-2) 8.2.1
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "arm-linux-gnueabihf".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from Cuberite...(no debugging symbols found)...done.
(gdb) set logging on
Copying output to gdb.txt.
(gdb) run
Starting program: /home/pi/Server/Cuberite
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/arm-linux-gnueabihf/libthread_db.so.1".
[New Thread 0xb6c34a80 (LWP 1100)]
[New Thread 0xb6433a80 (LWP 1101)]
[20:11:10] --- Started Log ---
[20:11:10] Cuberite local build (id: Unknown)
[20:11:10] from commit approx: abc96bb4ce0bfb5627e74f4f2e66f19eaaa12736 built at                                                                                                             : approx: 2020-12-24 16:26:23 +0000
[20:11:10] Creating new server instance...
[20:11:10] Reading server config...
[20:11:10] Starting server...
[New Thread 0xb5afea80 (LWP 1102)]
[20:11:10] Compatible clients: 1.8.x-1.12.x
[20:11:10] WebServer: The server will run in unsecured HTTP mode.
[20:11:10] Put a valid HTTPS certificate in file 'webadmin/httpscert.crt' and it                                                                                                             s corresponding private key to 'webadmin/httpskey.pem' (without any password) to                                                                                                              enable HTTPS support
[20:11:10] Loaded 789 crafting recipes
[20:11:10] Loaded 1.12.2 380 recipe book
[20:11:10] Cannot open file "Protocol/1.14.4/base.recipes.txt", no recipe book r                                                                                                             ecipes will be available!
[20:11:10] Cannot open file "Protocol/1.13/base.recipes.txt", no recipe book rec                                                                                                             ipes will be available!
[20:11:10] Loaded 67 furnace recipes and 66 fuels
[20:11:10] Loaded 49 brewing recipes
[20:11:10] Invalid biome ("BambooJungle") specified in AllowedBiomes in cubeset                                                                                                              file Prefabs/SinglePieceStructures/JungleTemple.cubeset. Skipping the biome.
[20:11:10] Invalid biome ("BambooJungleHills") specified in AllowedBiomes in cub                                                                                                             eset file Prefabs/SinglePieceStructures/JungleTemple.cubeset. Skipping the biome                                                                                                             .
[20:11:10] -- Loading Plugins --
[20:11:10] Initialised Core
[20:11:10] Initialized ChatLog v.3
[20:11:10] -- Loaded 2 Plugins --
[New Thread 0xb232ca80 (LWP 1103)]
[New Thread 0xb1b2ba80 (LWP 1104)]
[New Thread 0xb132aa80 (LWP 1105)]
[New Thread 0xb0b29a80 (LWP 1106)]
[New Thread 0xb0328a80 (LWP 1107)]
[New Thread 0xafb27a80 (LWP 1108)]
[20:11:11] Preparing spawn (world): 99.50% (398/400; 397.60 chunks / sec)
[New Thread 0xab1fea80 (LWP 1109)]
[New Thread 0xaa9fda80 (LWP 1110)]
[New Thread 0xaa1fca80 (LWP 1111)]
[New Thread 0xa99fba80 (LWP 1112)]
[New Thread 0xa91faa80 (LWP 1113)]
[20:11:13] Preparing spawn (world_nether): 54.50% (218/400; 199.09 chunks / sec)
[New Thread 0xa33fea80 (LWP 1114)]
[New Thread 0xa2bfda80 (LWP 1115)]
[New Thread 0xa23fca80 (LWP 1116)]
[New Thread 0xa1bfba80 (LWP 1117)]
[New Thread 0xa13faa80 (LWP 1118)]
[20:11:13] World "world_the_end": Did not find an acceptable spawnpoint. Generat                                                                                                             ed a random spawnpoint position at {0.00, 128.00, 0.00}
[New Thread 0x9fdfea80 (LWP 1119)]
[New Thread 0x9f5fda80 (LWP 1120)]
[20:11:13] WebAdmin is running on port(s) 8080
[20:11:13] Startup complete, took 3033ms!
[20:11:17] User BlockMuster_ authenticated with UUID 2cfaf5e2a81543a6aaff6e3660a                                                                                                             fc1ad
[20:11:17] Player BlockMuster_ has joined the game
[20:11:41] User Groffel authenticated with UUID 01d6bca9bcc54f3190cf2c22b033e3c6
[20:11:41] Player Groffel has joined the game
[20:12:35] User Luke_1310 authenticated with UUID a03d2603f63845cbb5c4ea76fdc885                                                                                                             b1
[20:12:35] Player Luke_1310 has joined the game
[20:15:18] MoveEntityToNewChunk: Failed to move entity, destination chunk unreac                                                                                                             hable. Entity lost
[20:26:51] MoveEntityToNewChunk: Failed to move entity, destination chunk unreac                                                                                                             hable. Entity lost
[20:29:29] MoveEntityToNewChunk: Failed to move entity, destination chunk unreac                                                                                                             hable. Entity lost
[20:30:19] MoveEntityToNewChunk: Failed to move entity, destination chunk unreac                                                                                                             hable. Entity lost
[20:32:42] MoveEntityToNewChunk: Failed to move entity, destination chunk unreac                                                                                                             hable. Entity lost
[20:32:52] MoveEntityToNewChunk: Failed to move entity, destination chunk unreac                                                                                                             hable. Entity lost
[20:36:42] MoveEntityToNewChunk: Failed to move entity, destination chunk unreac                                                                                                             hable. Entity lost
[20:39:19] MoveEntityToNewChunk: Failed to move entity, destination chunk unreac                                                                                                             hable. Entity lost
[20:46:53] User Easychat authenticated with UUID 96d5e6c1967f425aafbd117028708c9                                                                                                             a
[20:46:53] Player data file not found for Easychat (96d5e6c1967f425aafbd11702870                                                                                                             8c9a, offline e4322156c617372faff16206a009898d (unused)), will be reset to defau                                                                                                             lts.
[20:46:53] Player Easychat has joined the game
[20:47:08] Player Easychat has left the game
[20:47:33] Deadlock detected: world world has been stuck at age 451069. Aborting                                                                                                              the server.
[20:47:33] CS at 0xb4bad420, World world chunkmap: RecursionCount = 1, ThreadIDH                                                                                                             ash = dd89178c
[20:47:33] CS at 0xb52fe69c, World world clients: RecursionCount = 0, ThreadIDHa                                                                                                             sh = dd89178c
[20:47:33] CS at 0xb52fe670, World world tasks: RecursionCount = 0, ThreadIDHash                                                                                                              = dd89178c
[20:47:33] CS at 0xb3b8c420, World world_nether chunkmap: RecursionCount = 0, Th                                                                                                             readIDHash = 9797151
[20:47:33] CS at 0xb42dd69c, World world_nether clients: RecursionCount = 0, Thr                                                                                                             eadIDHash = 9797151
[20:47:33] CS at 0xb42dd670, World world_nether tasks: RecursionCount = 0, Threa                                                                                                             dIDHash = 9797151
[20:47:33] CS at 0xb2bb4420, World world_the_end chunkmap: RecursionCount = 0, T                                                                                                             hreadIDHash = 97a41ff4
[20:47:33] CS at 0xb330569c, World world_the_end clients: RecursionCount = 0, Th                                                                                                             readIDHash = 97a41ff4
[20:47:33] CS at 0xb3305670, World world_the_end tasks: RecursionCount = 0, Thre                                                                                                             adIDHash = 97a41ff4
[20:47:33] CS at 0x1995e20, cLuaState plugin Debuggers: RecursionCount = 0, Thre                                                                                                             adIDHash = 26b5ac16
[20:47:33] CS at 0x19812b8, cLuaState plugin ChatLog: RecursionCount = 0, Thread                                                                                                             IDHash = 382cca49
[20:47:33] CS at 0x2a07658, cLuaState plugin ProtectionAreas: RecursionCount = 0                                                                                                             , ThreadIDHash = 26b5ac16
[20:47:33] CS at 0x2a077a0, cLuaState plugin Core: RecursionCount = 0, ThreadIDH                                                                                                             ash = 97a41ff4
[20:47:33] CS at 0x197f648, cLuaState plugin HookNotify: RecursionCount = 0, Thr                                                                                                             eadIDHash = 26b5ac16
[20:47:33] CS at 0x197f7b0, cLuaState plugin TestLuaRocks: RecursionCount = 0, T                                                                                                             hreadIDHash = 26b5ac16
[20:47:33] CS at 0x197ef50, cLuaState plugin NetworkTest: RecursionCount = 0, Th                                                                                                             readIDHash = 26b5ac16
[20:47:33] CS at 0x197f098, cLuaState plugin DumpInfo: RecursionCount = 0, Threa                                                                                                             dIDHash = 26b5ac16
[20:47:33] CS at 0x197f1b8, cLuaState plugin APIDump: RecursionCount = 0, Thread                                                                                                             IDHash = 26b5ac16

Thread 21 "Cuberite" received signal SIGABRT, Aborted.
[Switching to Thread 0x9fdfea80 (LWP 1119)]
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50      ../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) thread apply all backtrace

Thread 22 (Thread 0x9f5fda80 (LWP 1120)):
#0  __lll_lock_wait (futex=futex@entry=0xb4bad428, private=0)
    at lowlevellock.c:43
#1  0xb6f84f58 in __GI___pthread_mutex_lock (mutex=0xb4bad428)
    at pthread_mutex_lock.c:115
#2  0x0011afc8 in cProtocol_1_8_0::HandlePacketLoginStart(cByteBuffer&) ()
#3  0x000f7948 in cProtocol_1_12_1::HandlePacket(cByteBuffer&, unsigned int) ()
#4  0x0010f4dc in cProtocol_1_8_0::AddReceivedData(cByteBuffer&, char const*, un                                                                                                             signed int) ()
#5  0x000ea4cc in cMultiVersionProtocol::HandleIncomingDataInRecognitionStage(cC                                                                                                             lientHandle&, std::basic_string_view<char, std::char_traits<char> >) ()
#6  0x00328470 in cClientHandle::ProcessProtocolInOut() ()
#7  0x002ff450 in cServer::cTickThread::Execute() ()
#8  0xb6ed29b0 in ?? () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#9  0xb6f82494 in start_thread (arg=0x9f5fda80) at pthread_create.c:486
#10 0xb6d0f578 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73
   from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 21 (Thread 0x9fdfea80 (LWP 1119)):
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0xb6c4f230 in __GI_abort () at abort.c:79
--Type <RET> for more, q to quit, c to continue without paging--
#2  0x00308a9c in cDeadlockDetect::DeadlockDetected(std::__cxx11::basic_string<c                                                                                                             har, std::char_traits<char>, std::allocator<char> > const&, long long) ()
#3  0x00308e14 in cDeadlockDetect::Execute() ()
#4  0xb6ed29b0 in ?? () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#5  0xb6f82494 in start_thread (arg=0x9fdfea80) at pthread_create.c:486
#6  0xb6d0f578 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73
   from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 20 (Thread 0xa13faa80 (LWP 1118)):
#0  __GI___nanosleep (remaining=0xa13fa3b8, requested_time=0xa13fa3b8)
    at ../sysdeps/unix/sysv/linux/nanosleep.c:28
#1  __GI___nanosleep (requested_time=0xa13fa3b8, remaining=0xa13fa3b8)
    at ../sysdeps/unix/sysv/linux/nanosleep.c:25
#2  0x002df568 in cWorld::cTickThread::Execute() ()
#3  0xb6ed29b0 in ?? () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#4  0xb6f82494 in start_thread (arg=0xa13faa80) at pthread_create.c:486
#5  0xb6d0f578 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73
   from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 19 (Thread 0xa1bfba80 (LWP 1117)):
#0  futex_wait_cancelable (private=0, expected=0, futex_word=0xb2bb4a20)
--Type <RET> for more, q to quit, c to continue without paging--
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88
#1  __pthread_cond_wait_common (abstime=0x0, mutex=0x0, cond=0xb2bb49f8)
    at pthread_cond_wait.c:502
#2  __pthread_cond_wait (cond=0xb2bb49f8, mutex=0x0) at pthread_cond_wait.c:655
#3  0xb6eccc94 in std::condition_variable::wait(std::unique_lock<std::mutex>&)
    () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#4  0x0012ff80 in cEvent::Wait() ()
#5  0x0033a6b0 in cChunkSender::Execute() ()
#6  0xb6ed29b0 in ?? () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#7  0xb6f82494 in start_thread (arg=0xa1bfba80) at pthread_create.c:486
#8  0xb6d0f578 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73
   from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 18 (Thread 0xa23fca80 (LWP 1116)):
#0  futex_wait_cancelable (private=0, expected=0, futex_word=0xb2bb4630)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88
#1  __pthread_cond_wait_common (abstime=0x0, mutex=0x0, cond=0xb2bb4608)
    at pthread_cond_wait.c:502
#2  __pthread_cond_wait (cond=0xb2bb4608, mutex=0x0) at pthread_cond_wait.c:655
#3  0xb6eccc94 in std::condition_variable::wait(std::unique_lock<std::mutex>&)
    () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#4  0x0012ff80 in cEvent::Wait() ()
--Type <RET> for more, q to quit, c to continue without paging--
#5  0x00350784 in cChunkGeneratorThread::Execute() ()
#6  0xb6ed29b0 in ?? () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#7  0xb6f82494 in start_thread (arg=0xa23fca80) at pthread_create.c:486
#8  0xb6d0f578 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73
   from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 17 (Thread 0xa2bfda80 (LWP 1115)):
#0  futex_wait_cancelable (private=0, expected=0, futex_word=0xb2bb4414)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88
#1  __pthread_cond_wait_common (abstime=0x0, mutex=0x0, cond=0xb2bb43e8)
    at pthread_cond_wait.c:502
#2  __pthread_cond_wait (cond=0xb2bb43e8, mutex=0x0) at pthread_cond_wait.c:655
#3  0xb6eccc94 in std::condition_variable::wait(std::unique_lock<std::mutex>&)
    () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#4  0x0012ff80 in cEvent::Wait() ()
#5  0x000ab2c0 in cWorldStorage::Execute() ()
#6  0xb6ed29b0 in ?? () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#7  0xb6f82494 in start_thread (arg=0xa2bfda80) at pthread_create.c:486
#8  0xb6d0f578 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73
   from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

--Type <RET> for more, q to quit, c to continue without paging--
Thread 16 (Thread 0xa33fea80 (LWP 1114)):
#0  futex_wait_cancelable (private=0, expected=0, futex_word=0xb2bb4c94)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88
#1  __pthread_cond_wait_common (abstime=0x0, mutex=0x0, cond=0xb2bb4c68)
    at pthread_cond_wait.c:502
#2  __pthread_cond_wait (cond=0xb2bb4c68, mutex=0x0) at pthread_cond_wait.c:655
#3  0xb6eccc94 in std::condition_variable::wait(std::unique_lock<std::mutex>&)
    () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#4  0x0012ff80 in cEvent::Wait() ()
#5  0x0031a6bc in cLightingThread::Execute() ()
#6  0xb6ed29b0 in ?? () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#7  0xb6f82494 in start_thread (arg=0xa33fea80) at pthread_create.c:486
#8  0xb6d0f578 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73
   from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 15 (Thread 0xa91faa80 (LWP 1113)):
#0  __GI___nanosleep (remaining=0xa91fa3b8, requested_time=0xa91fa3b8)
    at ../sysdeps/unix/sysv/linux/nanosleep.c:28
#1  __GI___nanosleep (requested_time=0xa91fa3b8, remaining=0xa91fa3b8)
    at ../sysdeps/unix/sysv/linux/nanosleep.c:25
#2  0x002df568 in cWorld::cTickThread::Execute() ()
#3  0xb6ed29b0 in ?? () from /lib/arm-linux-gnueabihf/libstdc++.so.6
--Type <RET> for more, q to quit, c to continue without paging--
#4  0xb6f82494 in start_thread (arg=0xa91faa80) at pthread_create.c:486
#5  0xb6d0f578 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73
   from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 14 (Thread 0xa99fba80 (LWP 1112)):
#0  futex_wait_cancelable (private=0, expected=0, futex_word=0xb3b8ca20)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88
#1  __pthread_cond_wait_common (abstime=0x0, mutex=0x0, cond=0xb3b8c9f8)
    at pthread_cond_wait.c:502
#2  __pthread_cond_wait (cond=0xb3b8c9f8, mutex=0x0) at pthread_cond_wait.c:655
#3  0xb6eccc94 in std::condition_variable::wait(std::unique_lock<std::mutex>&)
    () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#4  0x0012ff80 in cEvent::Wait() ()
#5  0x0033a6b0 in cChunkSender::Execute() ()
#6  0xb6ed29b0 in ?? () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#7  0xb6f82494 in start_thread (arg=0xa99fba80) at pthread_create.c:486
#8  0xb6d0f578 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73
   from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 13 (Thread 0xaa1fca80 (LWP 1111)):
#0  futex_wait_cancelable (private=0, expected=0, futex_word=0xb3b8c630)
--Type <RET> for more, q to quit, c to continue without paging--
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88
#1  __pthread_cond_wait_common (abstime=0x0, mutex=0x0, cond=0xb3b8c608)
    at pthread_cond_wait.c:502
#2  __pthread_cond_wait (cond=0xb3b8c608, mutex=0x0) at pthread_cond_wait.c:655
#3  0xb6eccc94 in std::condition_variable::wait(std::unique_lock<std::mutex>&)
    () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#4  0x0012ff80 in cEvent::Wait() ()
#5  0x00350784 in cChunkGeneratorThread::Execute() ()
#6  0xb6ed29b0 in ?? () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#7  0xb6f82494 in start_thread (arg=0xaa1fca80) at pthread_create.c:486
#8  0xb6d0f578 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73
   from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 12 (Thread 0xaa9fda80 (LWP 1110)):
#0  futex_wait_cancelable (private=0, expected=0, futex_word=0xb3b8c410)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88
#1  __pthread_cond_wait_common (abstime=0x0, mutex=0x0, cond=0xb3b8c3e8)
    at pthread_cond_wait.c:502
#2  __pthread_cond_wait (cond=0xb3b8c3e8, mutex=0x0) at pthread_cond_wait.c:655
#3  0xb6eccc94 in std::condition_variable::wait(std::unique_lock<std::mutex>&)
    () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#4  0x0012ff80 in cEvent::Wait() ()
--Type <RET> for more, q to quit, c to continue without paging--
#5  0x000ab2c0 in cWorldStorage::Execute() ()
#6  0xb6ed29b0 in ?? () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#7  0xb6f82494 in start_thread (arg=0xaa9fda80) at pthread_create.c:486
#8  0xb6d0f578 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73
   from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 11 (Thread 0xab1fea80 (LWP 1109)):
#0  futex_wait_cancelable (private=0, expected=0, futex_word=0xb3b8cc94)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88
#1  __pthread_cond_wait_common (abstime=0x0, mutex=0x0, cond=0xb3b8cc68)
    at pthread_cond_wait.c:502
#2  __pthread_cond_wait (cond=0xb3b8cc68, mutex=0x0) at pthread_cond_wait.c:655
#3  0xb6eccc94 in std::condition_variable::wait(std::unique_lock<std::mutex>&)
    () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#4  0x0012ff80 in cEvent::Wait() ()
#5  0x0031a6bc in cLightingThread::Execute() ()
#6  0xb6ed29b0 in ?? () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#7  0xb6f82494 in start_thread (arg=0xab1fea80) at pthread_create.c:486
#8  0xb6d0f578 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73
   from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

--Type <RET> for more, q to quit, c to continue without paging--
Thread 10 (Thread 0xafb27a80 (LWP 1108)):
#0  futex_wait_cancelable (private=0, expected=0, futex_word=0xb4bad4a0)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88
#1  __pthread_cond_wait_common (abstime=0x0, mutex=0x0, cond=0xb4bad478)
    at pthread_cond_wait.c:502
#2  __pthread_cond_wait (cond=0xb4bad478, mutex=0x0) at pthread_cond_wait.c:655
#3  0xb6eccc94 in std::condition_variable::wait(std::unique_lock<std::mutex>&)
    () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#4  0x0012ff80 in cEvent::Wait() ()
#5  0x00343730 in cChunkMap::GetHeight(int, int) ()
#6  0x002cecf4 in cWorld::IsWeatherWetAtXYZ(Vector3<int>) ()
#7  0x000d6504 in cFireSimulator::SimulateChunk(std::chrono::duration<long long,                                                                                                              std::ratio<1ll, 1000ll> >, int, int, cChunk*) ()
#8  0x002ddb88 in cWorld::cTickThread::Execute() ()
#9  0xb6ed29b0 in ?? () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#10 0xb6f82494 in start_thread (arg=0xafb27a80) at pthread_create.c:486
#11 0xb6d0f578 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73
   from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 9 (Thread 0xb0328a80 (LWP 1107)):
#0  __lll_lock_wait (futex=futex@entry=0xb4bad428, private=0)
    at lowlevellock.c:46
--Type <RET> for more, q to quit, c to continue without paging--
#1  0xb6f84f58 in __GI___pthread_mutex_lock (mutex=0xb4bad428)
    at pthread_mutex_lock.c:115
#2  0x0033cb90 in cChunkSender::Execute() ()
#3  0xb6ed29b0 in ?? () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#4  0xb6f82494 in start_thread (arg=0xb0328a80) at pthread_create.c:486
#5  0xb6d0f578 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73
   from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 8 (Thread 0xb0b29a80 (LWP 1106)):
#0  futex_wait_cancelable (private=0, expected=0, futex_word=0xb4bad634)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88
#1  __pthread_cond_wait_common (abstime=0x0, mutex=0x0, cond=0xb4bad608)
    at pthread_cond_wait.c:502
#2  __pthread_cond_wait (cond=0xb4bad608, mutex=0x0) at pthread_cond_wait.c:655
#3  0xb6eccc94 in std::condition_variable::wait(std::unique_lock<std::mutex>&)
    () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#4  0x0012ff80 in cEvent::Wait() ()
#5  0x00350784 in cChunkGeneratorThread::Execute() ()
#6  0xb6ed29b0 in ?? () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#7  0xb6f82494 in start_thread (arg=0xb0b29a80) at pthread_create.c:486
#8  0xb6d0f578 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73
   from /lib/arm-linux-gnueabihf/libc.so.6
--Type <RET> for more, q to quit, c to continue without paging--
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 7 (Thread 0xb132aa80 (LWP 1105)):
#0  futex_wait_cancelable (private=0, expected=0, futex_word=0xb4bad414)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88
#1  __pthread_cond_wait_common (abstime=0x0, mutex=0x0, cond=0xb4bad3e8)
    at pthread_cond_wait.c:502
#2  __pthread_cond_wait (cond=0xb4bad3e8, mutex=0x0) at pthread_cond_wait.c:655
#3  0xb6eccc94 in std::condition_variable::wait(std::unique_lock<std::mutex>&)
    () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#4  0x0012ff80 in cEvent::Wait() ()
#5  0x000ab2c0 in cWorldStorage::Execute() ()
#6  0xb6ed29b0 in ?? () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#7  0xb6f82494 in start_thread (arg=0xb132aa80) at pthread_create.c:486
#8  0xb6d0f578 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73
   from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 6 (Thread 0xb1b2ba80 (LWP 1104)):
#0  __lll_lock_wait (futex=futex@entry=0xb4bad428, private=0)
    at lowlevellock.c:43
#1  0xb6f84f58 in __GI___pthread_mutex_lock (mutex=0xb4bad428)
    at pthread_mutex_lock.c:115
--Type <RET> for more, q to quit, c to continue without paging--
#2  0x0031a218 in cLightingThread::LightChunk(cLightingThread::cLightingChunkSta                                                                                                             y&) [clone .part.42] ()
#3  0x0031a680 in cLightingThread::Execute() ()
#4  0xb6ed29b0 in ?? () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#5  0xb6f82494 in start_thread (arg=0xb1b2ba80) at pthread_create.c:486
#6  0xb6d0f578 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73
   from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 5 (Thread 0xb232ca80 (LWP 1103)):
#0  futex_wait_cancelable (private=0, expected=0, futex_word=0xbefff4c0)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88
#1  __pthread_cond_wait_common (abstime=0x0, mutex=0x0, cond=0xbefff498)
    at pthread_cond_wait.c:502
#2  __pthread_cond_wait (cond=0xbefff498, mutex=0x0) at pthread_cond_wait.c:655
#3  0xb6eccc94 in std::condition_variable::wait(std::unique_lock<std::mutex>&)
    () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#4  0x0012ff80 in cEvent::Wait() ()
#5  0x0011d64c in cAuthenticator::Execute() ()
#6  0xb6ed29b0 in ?? () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#7  0xb6f82494 in start_thread (arg=0xb232ca80) at pthread_create.c:486
#8  0xb6d0f578 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73
   from /lib/arm-linux-gnueabihf/libc.so.6
--Type <RET> for more, q to quit, c to continue without paging--
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 4 (Thread 0xb5afea80 (LWP 1102)):
#0  futex_abstimed_wait_cancelable (private=<optimized out>,
    abstime=0xb5afe448, expected=0, futex_word=0x575e38)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:205
#1  __pthread_cond_wait_common (abstime=0xb5afe448, mutex=0x0, cond=0x575e10)
    at pthread_cond_wait.c:539
#2  __pthread_cond_timedwait (cond=0x575e10, mutex=0x0, abstime=0xb5afe448)
    at pthread_cond_wait.c:667
#3  0x00120888 in cMojangAPI::cUpdateThread::Execute() ()
#4  0xb6ed29b0 in ?? () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#5  0xb6f82494 in start_thread (arg=0xb5afea80) at pthread_create.c:486
#6  0xb6d0f578 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73
   from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 3 (Thread 0xb6433a80 (LWP 1101)):
#0  0xb6d0f9d0 in epoll_wait (epfd=<optimized out>, events=0x572b08,
    maxevents=32, timeout=-1) at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
#1  0x00477b70 in epoll_dispatch ()
#2  0x0046c494 in event_base_loop ()
#3  0x00123f08 in cNetworkSingleton::RunEventLoop(cNetworkSingleton*) ()
--Type <RET> for more, q to quit, c to continue without paging--
#4  0xb6ed29b0 in ?? () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#5  0xb6f82494 in start_thread (arg=0xb6433a80) at pthread_create.c:486
#6  0xb6d0f578 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73
   from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 2 (Thread 0xb6c34a80 (LWP 1100)):
#0  futex_wait_cancelable (private=0, expected=0,
    futex_word=0x567830 <cNetworkSingleton::Get()::Instance+392>)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88
#1  __pthread_cond_wait_common (abstime=0x0, mutex=0x0,
    cond=0x567808 <cNetworkSingleton::Get()::Instance+352>)
    at pthread_cond_wait.c:502
#2  __pthread_cond_wait (
    cond=0x567808 <cNetworkSingleton::Get()::Instance+352>, mutex=0x0)
    at pthread_cond_wait.c:655
#3  0xb6eccc94 in std::condition_variable::wait(std::unique_lock<std::mutex>&)
    () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#4  0x0012ff80 in cEvent::Wait() ()
#5  0x001300a0 in cNetworkLookup::Execute() ()
#6  0xb6ed29b0 in ?? () from /lib/arm-linux-gnueabihf/libstdc++.so.6
#7  0xb6f82494 in start_thread (arg=0xb6c34a80) at pthread_create.c:486
#8  0xb6d0f578 in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73
--Type <RET> for more, q to quit, c to continue without paging--
   from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 1 (Thread 0xb6fef410 (LWP 1097)):
#0  __GI___select (timeout=0x0, exceptfds=0x0, writefds=0x0,
    readfds=0xbeffe9f0, nfds=1) at ../sysdeps/unix/sysv/linux/select.c:41
#1  __GI___select (nfds=1, readfds=0xbeffe9f0, writefds=0x0, exceptfds=0x0,
    timeout=0xbeffe8c8) at ../sysdeps/unix/sysv/linux/select.c:37
#2  0x0004423c in cRoot::Run(cSettingsRepositoryInterface&) [clone .constprop.11                                                                                                             63] ()
#3  0x000489b8 in UniversalMain(int, char**, bool) [clone .constprop.1003] ()
#4  0xb6c4f718 in __libc_start_main (main=0xbefff6d4, argc=-1227350016,
    argv=0xb6c4f718 <__libc_start_main+268>, init=<optimized out>,
    fini=0x479bf4 <__libc_csu_fini>, rtld_fini=0xb6fde4c4 <_dl_fini>,
    stack_end=0xbefff6d4) at libc-start.c:308
#5  0x00018010 in _start ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
(gdb)
(gdb) generate-core-file dump.core
Saved corefile dump.core
(gdb) quit
A debugging session is active.

        Inferior 1 [process 1097] will be killed.

Quit anyway? (y or n) y
pi@raspberrypi:~/Server $ gdb.txt
-bash: gdb.txt: command not found

so last time i downloaded your githubfile, but then my internet crashes, and i need to do it again. 
But now, the cmake will not work any longer...


pi@raspberrypi:~ $ cd
pi@raspberrypi:~ $ mv cuberite cuberite-old
mv: cannot move 'cuberite' to 'cuberite-old/cuberite': Directory not empty
pi@raspberrypi:~ $ git clone https://github.com/cuberite/cuberite.git
fatal: destination path 'cuberite' already exists and is not an empty directory.
pi@raspberrypi:~ $ cd cuberite
pi@raspberrypi:~/cuberite $ git checkout Deadlock
Already on 'Deadlock'
Your branch is up to date with 'origin/Deadlock'.
pi@raspberrypi:~/cuberite $ mkdir Build
mkdir: cannot create directory ‘Build’: File exists
pi@raspberrypi:~/cuberite $ cd Build
pi@raspberrypi:~/cuberite/Build $ cmake .. -DCMAKE_BUILD_TYPE=RelWithDebInfo
CMake Error at CMake/AddDependencies.cmake:35 (message):
  expat is missing in folder lib/expat.  Have you initialized and updated the
  submodules / downloaded the extra libraries?
Call Stack (most recent call first):
  CMakeLists.txt:65 (build_dependencies)


-- Configuring incomplete, errors occurred!
See also "/home/pi/cuberite/Build/CMakeFiles/CMakeOutput.log".

--- a/src/Blocks/BlockFarmland.h
+++ b/src/Blocks/BlockFarmland.h
@@ -115,12 +115,13 @@ private:
 	/** Returns true if there's either a water source block close enough to hydrate the specified position, or it's raining there. */
 	static bool IsWaterInNear(const cChunk & a_Chunk, const Vector3i a_RelPos)
 	{
-		const auto WorldPos = a_Chunk.RelativeToAbsolute(a_RelPos);
-		if (a_Chunk.GetWorld()->IsWeatherWetAtXYZ(WorldPos))
+		if (a_Chunk.IsWeatherWetAt(a_RelPos.addedY(1)))
 		{
 			// Rain hydrates farmland, too
 			return true;
 		}
+
+		const auto WorldPos = a_Chunk.RelativeToAbsolute(a_RelPos);
 
 		// Search for water in a close proximity:
 		// Ref.: https://minecraft.gamepedia.com/Farmland#Hydration

--- a/src/Chunk.cpp
+++ b/src/Chunk.cpp
@@ -627,7 +627,6 @@ void cChunk::SpawnMobs(cMobSpawner & a_MobSpawner)
 		int WorldX, WorldY, WorldZ;
 		PositionToWorldPosition(TryX, TryY, TryZ, WorldX, WorldY, WorldZ);
 
-		EMCSBiome Biome = m_ChunkMap->GetBiomeAt(WorldX, WorldZ);
 		// MG TODO :
 		// Moon cycle (for slime)
 		// check player and playerspawn presence < 24 blocks
@@ -642,12 +641,16 @@ void cChunk::SpawnMobs(cMobSpawner & a_MobSpawner)
 		*/
 
 		NumberOfTries++;
-		if (!IsLightValid())
+
+		Vector3i Try(TryX, TryY, TryZ);
+		const auto Chunk = GetRelNeighborChunkAdjustCoords(Try);
+
+		if ((Chunk == nullptr) || !Chunk->IsValid() || !Chunk->IsLightValid())
 		{
 			continue;
 		}
 
-		auto newMob = a_MobSpawner.TryToSpawnHere(this, {TryX, TryY, TryZ}, Biome, MaxNbOfSuccess);
+		auto newMob = a_MobSpawner.TryToSpawnHere(this, Try, GetBiomeAt(Try.x, Try.z), MaxNbOfSuccess);
 		if (newMob == nullptr)
 		{
 			continue;
@@ -1191,15 +1194,56 @@ bool cChunk::UnboundedRelFastSetBlock(Vector3i a_RelPos, BLOCKTYPE a_BlockType, 
 
 
 
-int cChunk::GetHeight(int a_X, int a_Z)
+int cChunk::GetHeight(int a_X, int a_Z) const
 {
 	ASSERT((a_X >= 0) && (a_X < Width) && (a_Z >= 0) && (a_Z < Width));
+	return m_HeightMap[a_X + a_Z * Width];
+}
 
-	if ((a_X >= 0) && (a_X < Width) && (a_Z >= 0) && (a_Z < Width))
+
+
+
+
+bool cChunk::IsWeatherSunnyAt(int a_RelX, int a_RelZ) const
+{
+	return m_World->IsWeatherSunny() || IsBiomeNoDownfall(GetBiomeAt(a_RelX, a_RelZ));
+}
+
+
+
+
+
+bool cChunk::IsWeatherWetAt(const int a_RelX, const int a_RelZ) const
+{
+	const auto Biome = GetBiomeAt(a_RelX, a_RelZ);
+	return m_World->IsWeatherWet() && !IsBiomeNoDownfall(Biome) && !IsBiomeCold(Biome);
+}
+
+
+
+
+
+bool cChunk::IsWeatherWetAt(const Vector3i a_Position) const
+{
+	if ((a_Position.y < 0) || !IsWeatherWetAt(a_Position.x, a_Position.z))
 	{
-		return m_HeightMap[a_X + a_Z * Width];
+		return false;
 	}
-	return 0;
+
+	if (a_Position.y >= cChunkDef::Height)
+	{
+		return true;
+	}
+
+	for (int y = GetHeight(a_Position.x, a_Position.z); y >= a_Position.y; y--)
+	{
+		if (cBlockInfo::IsRainBlocker(GetBlock({ a_Position.x, y, a_Position.z })))
+		{
+			return false;
+		}
+	}
+
+	return true;
 }
 
 

--- a/src/Chunk.h
+++ b/src/Chunk.h
@@ -221,7 +221,17 @@ public:
 	/** Sets the sign text. Returns true if successful. Also sends update packets to all clients in the chunk */
 	bool SetSignLines(int a_RelX, int a_RelY, int a_RelZ, const AString & a_Line1, const AString & a_Line2, const AString & a_Line3, const AString & a_Line4);
 
-	int  GetHeight( int a_X, int a_Z);
+	int GetHeight( int a_X, int a_Z) const;
+
+	/** Returns true if it is sunny at the specified location. This takes into account biomes. */
+	bool IsWeatherSunnyAt(int a_RelX, int a_RelZ) const;
+
+	/** Returns true if it is raining or storming at the specified location, taking into account biomes. */
+	bool IsWeatherWetAt(int a_RelX, int a_RelZ) const;
+
+	/** Returns true if it is raining or storming at the specified location,
+	and the rain reaches (the bottom of) the specified block position. */
+	bool IsWeatherWetAt(Vector3i a_Position) const;
 
 	void SendBlockTo(int a_RelX, int a_RelY, int a_RelZ, cClientHandle * a_Client);
 

--- a/src/ChunkMap.cpp
+++ b/src/ChunkMap.cpp
@@ -340,6 +340,63 @@ bool cChunkMap::IsChunkQueued(int a_ChunkX, int a_ChunkZ) const
 
 
 
+bool cChunkMap::IsWeatherSunnyAt(int a_BlockX, int a_BlockZ) const
+{
+	int ChunkX, ChunkZ, BlockY = 0;
+	cChunkDef::AbsoluteToRelative(a_BlockX, BlockY, a_BlockZ, ChunkX, ChunkZ);
+
+	cCSLock Lock(m_CSChunks);
+	const auto Chunk = FindChunk(ChunkX, ChunkZ);
+	if ((Chunk == nullptr) || !Chunk->IsValid())
+	{
+		return m_World->IsWeatherSunny();
+	}
+
+	return Chunk->IsWeatherSunnyAt(a_BlockX, a_BlockZ);
+}
+
+
+
+
+
+bool cChunkMap::IsWeatherWetAt(int a_BlockX, int a_BlockZ) const
+{
+	int ChunkX, ChunkZ, BlockY = 0;
+	cChunkDef::AbsoluteToRelative(a_BlockX, BlockY, a_BlockZ, ChunkX, ChunkZ);
+
+	cCSLock Lock(m_CSChunks);
+	const auto Chunk = FindChunk(ChunkX, ChunkZ);
+	if ((Chunk == nullptr) || !Chunk->IsValid())
+	{
+		return m_World->IsWeatherWet();
+	}
+
+	return Chunk->IsWeatherWetAt(a_BlockX, a_BlockZ);
+}
+
+
+
+
+
+bool cChunkMap::IsWeatherWetAt(const Vector3i a_Position) const
+{
+	const auto ChunkPosition = cChunkDef::BlockToChunk(a_Position);
+	const auto Position = cChunkDef::AbsoluteToRelative(a_Position, ChunkPosition);
+
+	cCSLock Lock(m_CSChunks);
+	const auto Chunk = FindChunk(ChunkPosition.m_ChunkX, ChunkPosition.m_ChunkZ);
+	if ((Chunk == nullptr) || !Chunk->IsValid())
+	{
+		return m_World->IsWeatherWet();
+	}
+
+	return Chunk->IsWeatherWetAt(Position);
+}
+
+
+
+
+
 bool cChunkMap::IsChunkValid(int a_ChunkX, int a_ChunkZ) const
 {
 	cCSLock Lock(m_CSChunks);
@@ -650,10 +707,7 @@ EMCSBiome cChunkMap::GetBiomeAt(int a_BlockX, int a_BlockZ) const
 	{
 		return Chunk->GetBiomeAt(X, Z);
 	}
-	else
-	{
-		return m_World->GetGenerator().GetBiomeAt(a_BlockX, a_BlockZ);
-	}
+	return EMCSBiome::biInvalidBiome;
 }
 
 

--- a/src/ChunkMap.h
+++ b/src/ChunkMap.h
@@ -5,9 +5,6 @@
 
 #pragma once
 
-
-#include <functional>
-
 #include "ChunkDataCallback.h"
 #include "EffectID.h"
 #include "FunctionRef.h"
@@ -120,6 +117,10 @@ public:
 	/** Returns true iff the chunk is in the loader / generator queue. */
 	bool IsChunkQueued(int a_ChunkX, int a_ChunkZ) const;
 
+	bool IsWeatherSunnyAt(int a_BlockX, int a_BlockZ) const;
+	bool IsWeatherWetAt(int a_BlockX, int a_BlockZ) const;
+	bool IsWeatherWetAt(Vector3i a_Position) const;
+
 	bool      IsChunkValid       (int a_ChunkX, int a_ChunkZ) const;
 	bool      HasChunkAnyClients (int a_ChunkX, int a_ChunkZ) const;
 	int       GetHeight          (int a_BlockX, int a_BlockZ);  // Waits for the chunk to get loaded / generated
@@ -143,15 +144,15 @@ public:
 	Ignored if the chunk is invalid. */
 	void SetBlockMeta(Vector3i a_BlockPos, NIBBLETYPE a_BlockMeta);
 
-	void       SetBlock          (Vector3i a_BlockPos, BLOCKTYPE a_BlockType, NIBBLETYPE a_BlockMeta);
-	bool       GetBlockTypeMeta  (Vector3i a_BlockPos, BLOCKTYPE & a_BlockType, NIBBLETYPE & a_BlockMeta) const;
-	bool       GetBlockInfo      (Vector3i, BLOCKTYPE & a_BlockType, NIBBLETYPE & a_Meta, NIBBLETYPE & a_SkyLight, NIBBLETYPE & a_BlockLight) const;
+	void SetBlock          (Vector3i a_BlockPos, BLOCKTYPE a_BlockType, NIBBLETYPE a_BlockMeta);
+	bool GetBlockTypeMeta  (Vector3i a_BlockPos, BLOCKTYPE & a_BlockType, NIBBLETYPE & a_BlockMeta) const;
+	bool GetBlockInfo      (Vector3i, BLOCKTYPE & a_BlockType, NIBBLETYPE & a_Meta, NIBBLETYPE & a_SkyLight, NIBBLETYPE & a_BlockLight) const;
 
 	/** Special function used for growing trees, replaces only blocks that tree may overwrite */
-	void      ReplaceTreeBlocks(const sSetBlockVector & a_Blocks);
+	void ReplaceTreeBlocks(const sSetBlockVector & a_Blocks);
 
-	/** Returns the biome at the specified coords. Reads the biome from the chunk, if loaded, otherwise uses the world generator to provide the biome value */
-	EMCSBiome GetBiomeAt (int a_BlockX, int a_BlockZ) const;
+	/** Returns the biome at the specified coords. Reads the biome from the chunk, if loaded, invalid otherwise. */
+	EMCSBiome GetBiomeAt(int a_BlockX, int a_BlockZ) const;
 
 	/** Sets the biome at the specified coords. Returns true if successful, false if not (chunk not loaded).
 	Doesn't resend the chunk to clients. */

--- a/src/Entities/Entity.cpp
+++ b/src/Entities/Entity.cpp
@@ -1198,8 +1198,8 @@ void cEntity::TickBurning(cChunk & a_Chunk)
 		m_TicksLeftBurning = 0;
 	}
 
-	// Fire is extinguished by rain
-	if (GetWorld()->IsWeatherWetAtXYZ(GetPosition().Floor()))
+	// Fire is extinguished by rain:
+	if (a_Chunk.IsWeatherWetAt(cChunkDef::AbsoluteToRelative(GetPosition().Floor(), a_Chunk.GetPos())))
 	{
 		m_TicksLeftBurning = 0;
 	}

--- a/src/Entities/Floater.cpp
+++ b/src/Entities/Floater.cpp
@@ -101,11 +101,16 @@ void cFloater::SpawnOn(cClientHandle & a_Client)
 
 void cFloater::Tick(std::chrono::milliseconds a_Dt, cChunk & a_Chunk)
 {
-	auto & Random = GetRandomProvider();
-
 	HandlePhysics(a_Dt, a_Chunk);
-	if (IsBlockWater(m_World->GetBlock(POSX_TOINT, POSY_TOINT, POSZ_TOINT))
-		&& (m_World->GetBlockMeta(POSX_TOINT, POSY_TOINT, POSZ_TOINT) == 0))
+
+	PREPARE_REL_AND_CHUNK(GetPosition().Floor(), a_Chunk);
+	if (!RelSuccess)
+	{
+		return;
+	}
+
+	auto & Random = GetRandomProvider();
+	if (IsBlockWater(Chunk->GetBlock(Rel)) && (Chunk->GetMeta(Rel) == 0))
 	{
 		if (!m_CanPickupItem && (m_AttachedMobID == cEntity::INVALID_ID))  // Check if you can't already pickup a fish and if the floater isn't attached to a mob.
 		{
@@ -113,7 +118,7 @@ void cFloater::Tick(std::chrono::milliseconds a_Dt, cChunk & a_Chunk)
 			{
 				m_BitePos = GetPosition();
 				m_World->BroadcastSoundEffect("entity.bobber.splash", GetPosition(), 1, 1);
-				SetPosY(GetPosY() - 1);
+				AddSpeedY(-10);
 				m_CanPickupItem = true;
 				m_PickupCountDown = 20;
 				m_CountDownTime = Random.RandInt(100, 900);
@@ -132,9 +137,9 @@ void cFloater::Tick(std::chrono::milliseconds a_Dt, cChunk & a_Chunk)
 			}
 
 			m_CountDownTime--;
-			if (m_World->GetHeight(POSX_TOINT, POSZ_TOINT) == POSY_TOINT)
+			if (Chunk->IsWeatherWetAt(Rel))
 			{
-				if (m_World->IsWeatherWet() && Random.RandBool(0.25))  // 25% chance of an extra countdown when being rained on.
+				if (Random.RandBool(0.25))  // 25% chance of an extra countdown when being rained on.
 				{
 					m_CountDownTime--;
 				}
@@ -150,7 +155,10 @@ void cFloater::Tick(std::chrono::milliseconds a_Dt, cChunk & a_Chunk)
 	}
 
 	// Check water at the top of floater otherwise it floats into the air above the water
-	if (IsBlockWater(m_World->GetBlock(POSX_TOINT, FloorC(GetPosY() + GetHeight()), POSZ_TOINT)))
+	if (
+		const auto Above = Rel.addedY(FloorC(GetPosY() + GetHeight()));
+		(Above.y < cChunkDef::Height) && IsBlockWater(m_World->GetBlock(Above))
+	)
 	{
 		SetSpeedY(0.7);
 	}

--- a/src/MobSpawner.cpp
+++ b/src/MobSpawner.cpp
@@ -74,10 +74,6 @@ eMonsterType cMobSpawner::ChooseMobType(EMCSBiome a_Biome)
 
 bool cMobSpawner::CanSpawnHere(cChunk * a_Chunk, Vector3i a_RelPos, eMonsterType a_MobType, EMCSBiome a_Biome, bool a_DisableSolidBelowCheck)
 {
-	if (a_Chunk == nullptr)
-	{
-		return false;
-	}
 	if ((a_RelPos.y >= cChunkDef::Height - 1) || (a_RelPos.y <= 0))
 	{
 		return false;
@@ -501,9 +497,6 @@ cMonster * cMobSpawner::TryToSpawnHere(cChunk * a_Chunk, Vector3i a_RelPos, EMCS
 		}
 		m_NewPack = false;
 	}
-
-	// Make sure we are looking at the right chunk to spawn in
-	a_Chunk = a_Chunk->GetRelNeighborChunkAdjustCoords(a_RelPos);
 
 	if ((m_AllowedTypes.find(m_MobType) != m_AllowedTypes.end()) && CanSpawnHere(a_Chunk, a_RelPos, m_MobType, a_Biome))
 	{

--- a/src/Mobs/Enderman.cpp
+++ b/src/Mobs/Enderman.cpp
@@ -1,6 +1,7 @@
 
 #include "Globals.h"  // NOTE: MSVC stupidness requires this to be the same across all modules
 
+#include "Chunk.h"
 #include "Enderman.h"
 #include "../Entities/Player.h"
 #include "../LineBlockTracer.h"
@@ -148,11 +149,14 @@ void cEnderman::Tick(std::chrono::milliseconds a_Dt, cChunk & a_Chunk)
 		return;
 	}
 
-	// Take damage when wet
-	if (
-		cChunkDef::IsValidHeight(POSY_TOINT) &&
-		(GetWorld()->IsWeatherWetAtXYZ(GetPosition().Floor()) || IsInWater())
-	)
+	PREPARE_REL_AND_CHUNK(GetPosition().Floor(), a_Chunk);
+	if (!RelSuccess)
+	{
+		return;
+	}
+
+	// Take damage when wet:
+	if (IsInWater() || Chunk->IsWeatherWetAt(Rel))
 	{
 		EventLosePlayer();
 		TakeDamage(dtEnvironment, nullptr, 1, 0);

--- a/src/Mobs/Monster.cpp
+++ b/src/Mobs/Monster.cpp
@@ -1667,10 +1667,10 @@ bool cMonster::WouldBurnAt(Vector3d a_Location, cChunk & a_Chunk)
 	}
 
 	if (
-		(Chunk->GetBlock(Rel.x, Rel.y, Rel.z) != E_BLOCK_SOULSAND) &&  // Not on soulsand
-		(GetWorld()->GetTimeOfDay() < 12000 + 1000) &&              // Daytime
-		GetWorld()->IsWeatherSunnyAt(POSX_TOINT, POSZ_TOINT) &&     // Not raining
-		!IsInWater()                                                // Isn't swimming
+		(Chunk->GetBlock(Rel) != E_BLOCK_SOULSAND) &&   // Not on soulsand
+		(GetWorld()->GetTimeOfDay() < 12000 + 1000) &&  // Daytime
+		Chunk->IsWeatherSunnyAt(Rel.x, Rel.z) &&        // Not raining
+		!IsInWater()                                    // Isn't swimming
 	)
 	{
 		int MobHeight = CeilC(a_Location.y + GetHeight()) - 1;  // The block Y coord of the mob's head

--- a/src/Mobs/SnowGolem.cpp
+++ b/src/Mobs/SnowGolem.cpp
@@ -1,6 +1,7 @@
 
 #include "Globals.h"  // NOTE: MSVC stupidness requires this to be the same across all modules
 
+#include "Chunk.h"
 #include "SnowGolem.h"
 #include "../BlockInfo.h"
 #include "../World.h"
@@ -36,17 +37,22 @@ void cSnowGolem::Tick(std::chrono::milliseconds a_Dt, cChunk & a_Chunk)
 		// The base class tick destroyed us
 		return;
 	}
-	if (IsBiomeNoDownfall(m_World->GetBiomeAt(POSX_TOINT, POSZ_TOINT)))
+
+	PREPARE_REL_AND_CHUNK(GetPosition().Floor(), a_Chunk);
+	if (!RelSuccess)
+	{
+		return;
+	}
+
+	if (IsBiomeNoDownfall(Chunk->GetBiomeAt(Rel.x, Rel.z)))
 	{
 		TakeDamage(dtEnvironment, nullptr, GetRawDamageAgainst(*this), GetKnockbackAmountAgainst(*this));
 	}
-	else
+	else if (const auto Below = Rel.addedY(-1); Below.y >= 0)
 	{
-		BLOCKTYPE BlockBelow = m_World->GetBlock(POSX_TOINT, POSY_TOINT - 1, POSZ_TOINT);
-		BLOCKTYPE Block = m_World->GetBlock(POSX_TOINT, POSY_TOINT, POSZ_TOINT);
-		if ((Block == E_BLOCK_AIR) && cBlockInfo::IsSolid(BlockBelow))
+		if ((Chunk->GetBlock(Rel) == E_BLOCK_AIR) && cBlockInfo::IsSolid(Chunk->GetBlock(Below)))
 		{
-			m_World->SetBlock(POSX_TOINT, POSY_TOINT, POSZ_TOINT, E_BLOCK_SNOW, 0);
+			Chunk->SetBlock(Rel, E_BLOCK_SNOW, 0);
 		}
 	}
 }

--- a/src/Simulator/FireSimulator.cpp
+++ b/src/Simulator/FireSimulator.cpp
@@ -87,7 +87,6 @@ void cFireSimulator::SimulateChunk(std::chrono::milliseconds a_Dt, int a_ChunkX,
 	for (cCoordWithIntList::iterator itr = Data.begin(); itr != Data.end();)
 	{
 		Vector3i relPos(itr->x, itr->y, itr->z);
-		auto absPos = a_Chunk->RelativeToAbsolute(relPos);
 		auto blockType = a_Chunk->GetBlock(relPos);
 
 		if (!IsAllowedBlock(blockType))
@@ -101,12 +100,16 @@ void cFireSimulator::SimulateChunk(std::chrono::milliseconds a_Dt, int a_ChunkX,
 		auto BurnsForever = ((relPos.y > 0) && DoesBurnForever(a_Chunk->GetBlock(relPos.addedY(-1))));
 		auto BlockMeta = a_Chunk->GetMeta(relPos);
 
-		auto Raining = std::any_of(std::begin(gCrossCoords), std::end(gCrossCoords),
-			[this, absPos](Vector3i cc)
+		auto Raining = std::any_of(std::begin(gCrossCoords), std::end(gCrossCoords), [a_Chunk, relPos](Vector3i cc)
+		{
+			auto Adjusted = relPos + cc;
+			const auto Chunk = a_Chunk->GetRelNeighborChunkAdjustCoords(Adjusted);
+			if ((Chunk != nullptr) && Chunk->IsValid())
 			{
-				return (m_World.IsWeatherWetAtXYZ(absPos + cc));
+				return Chunk->IsWeatherWetAt(Adjusted);
 			}
-		);
+			return false;
+		});
 
 		// Randomly burn out the fire if it is raining:
 		if (!BurnsForever && Raining && GetRandomProvider().RandBool(CHANCE_BASE_RAIN_EXTINGUISH + (BlockMeta * CHANCE_AGE_M_RAIN_EXTINGUISH)))

--- a/src/World.cpp
+++ b/src/World.cpp
@@ -548,28 +548,27 @@ void cWorld::ChangeWeather(void)
 
 
 
-bool cWorld::IsWeatherWetAtXYZ(Vector3i a_Pos)
+bool cWorld::IsWeatherSunnyAt(int a_BlockX, int a_BlockZ) const
 {
-	if ((a_Pos.y < 0) || !IsWeatherWetAt(a_Pos.x, a_Pos.z))
-	{
-		return false;
-	}
+	return m_ChunkMap.IsWeatherSunnyAt(a_BlockX, a_BlockZ);
+}
 
-	if (a_Pos.y >= cChunkDef::Height)
-	{
-		return true;
-	}
 
-	for (int y = GetHeight(a_Pos.x, a_Pos.z); y >= a_Pos.y; y--)
-	{
-		auto BlockType = GetBlock({a_Pos.x, y, a_Pos.z});
-		if (cBlockInfo::IsRainBlocker(BlockType))
-		{
-			return false;
-		}
-	}
 
-	return true;
+
+
+bool cWorld::IsWeatherWetAt(int a_BlockX, int a_BlockZ)
+{
+	return m_ChunkMap.IsWeatherWetAt(a_BlockX, a_BlockZ);
+}
+
+
+
+
+
+bool cWorld::IsWeatherWetAtXYZ(const Vector3i a_Position)
+{
+	return m_ChunkMap.IsWeatherWetAt(a_Position);
 }
 
 

--- a/src/World.h
+++ b/src/World.h
@@ -999,10 +999,7 @@ public:
 	bool IsWeatherSunny(void) const { return (m_Weather == wSunny); }
 
 	/** Returns true if it is sunny at the specified location. This takes into account biomes. */
-	bool IsWeatherSunnyAt(int a_BlockX, int a_BlockZ)
-	{
-		return (IsWeatherSunny() || IsBiomeNoDownfall(GetBiomeAt(a_BlockX, a_BlockZ)));
-	}
+	bool IsWeatherSunnyAt(int a_BlockX, int a_BlockZ) const;
 
 	/** Returns true if the current weather is rainy. */
 	bool IsWeatherRain(void) const { return (m_Weather == wRain); }
@@ -1027,15 +1024,11 @@ public:
 
 	/** Returns true if it is raining or storming at the specified location.
 	This takes into account biomes. */
-	virtual bool IsWeatherWetAt(int a_BlockX, int a_BlockZ) override
-	{
-		auto Biome = GetBiomeAt(a_BlockX, a_BlockZ);
-		return (IsWeatherWet() && !IsBiomeNoDownfall(Biome) && !IsBiomeCold(Biome));
-	}
+	virtual bool IsWeatherWetAt(int a_BlockX, int a_BlockZ) override;
 
 	/** Returns true if it is raining or storming at the specified location,
 	and the rain reaches (the bottom of) the specified block position. */
-	virtual bool IsWeatherWetAtXYZ(Vector3i a_Pos) override;
+	virtual bool IsWeatherWetAtXYZ(Vector3i a_Position) override;
 
 	/** Returns the seed of the world. */
 	int GetSeed(void) { return m_Generator.GetSeed(); }

